src/frequenz/sdk/timeseries/_formula_engine/_formula_printer.py

```diff
@@ -0,0 +1,220 @@
+            case "-":
+                return OperatorPrecedence.Subtraction
+            case "*":
+                return OperatorPrecedence.Multiplication
+            case "/":
+                return OperatorPrecedence.Division
+            case _:
+                return OperatorPrecedence.Primary
+
+
+class StackItem:
+    """Represents a stack item for the formula printer."""
+
+    def __init__(self, value: str, precedence: OperatorPrecedence):
+        """Initialize the StackItem.
+
+        Args:
+            value: The value of the stack item.
+            precedence: The precedence of the stack item.
+        """
+        self.value = value
+        self.precedence = precedence
+
+    def __str__(self) -> str:
+        """Return the string representation of the stack item.
+
+        This is used for debugging purposes.
+
+        Returns:
+            str: The string representation of the stack item.
+        """
+        return f"(\"{self.value}\", {self.precedence})"
+
+    def plvalue(self, outer_precedence: OperatorPrecedence) -> str:
+        """Return the value of the stack item with parentheses if necessary.
+
+        Args:
+            outer_precedence: The precedence of the outer stack item.
+
+        Returns:
+            str: The value of the stack item with parentheses if necessary.
+        """
+        return f"({self.value})" if self.precedence < outer_precedence else self.value
+
+    def prvalue(self, outer_precedence: OperatorPrecedence) -> str:
+        """Return the value of the stack item with parentheses if necessary.
+
+        Args:
+            outer_precedence: The precedence of the outer stack item.
+
+        Returns:
+            str: The value of the stack item with parentheses if necessary.
+        """
+        return f"({self.value})" if self.precedence < outer_precedence else self.value
+
+    @staticmethod
+    def create_binary(lhs: StackItem, op: str, rhs: StackItem) -> StackItem:
+        """Create a binary stack item.
+
+        Args:
+            lhs: The left-hand side of the binary operation.
+            op: The operator of the binary operation.
+            rhs: The right-hand side of the binary operation.
+
+        Returns:
+            StackItem: The binary stack item.
+        """
+        pred = OperatorPrecedence.map(op)
+        return StackItem(
+            f"{lhs.plvalue(pred)} {op} {rhs.prvalue(pred)}",
+            pred,
+        )
+
+    @staticmethod
+    def create_primary(value: float) -> StackItem:
+        """Create a stack item for literal values or function calls (primary expressions).
+
+        Args:
+            value: The value of the literal.
+
+        Returns:
+            StackItem: The literal stack item.
+        """
+        return StackItem(str(value), OperatorPrecedence.Primary)
+
+
+class FormulaPrinter:
+    """Prints a formula."""
+
+    def __init__(self) -> None:
+        """Initialize the FormulaPrinter."""
+        self._stack = list[StackItem]()
```

Comment:
I guess this will be small enough to not matter, but it is more suitable to use a `collections.deque` as a stack instead of a `list`.

Reply:
I think it does not matter, because we neither need index-access, nor do we manipulate the front of the stack. What do you think?

Reply:
Confusingly enough, `list` is actually a C array in Python, so needing indexing is actually a reason to use `list`, not to use it :P

A `deque` is a list of arrays, so indexing is potentially O(n), but pushing and popping from any end is amortized constant time.

Again, if we are talking about less than hundreds or items, probably `list` and `deque` are exactly the same. So it was more of an informational comment, because I guess we'll be way below that.

All of this IIRC.

Reply:
Python is sooo confusing.
However, I do not need any indexing. I just extend at the end (which ever end, just happens to be the right end), and pop off at at that. I expect Python's `list` implementation to be intelligent enough to not resize upon every append and pop, but instead having an internal `used_elements` counter that tells how many elements are in that list, effectively making a `pop()` O(1) all the times (not so good for first pushes). But I do not see how such kind of optimization is actually useful in this context, because this functionality is most likely only being used for debugging / logging / introspection, and least likely in production hot paths.